clang/test/SemaTemplate/instantiate-requires-expr.cpp

```diff
@@ -36,7 +36,7 @@ using r1i2 = r1<char>; // expected-error {{constraints not satisfied for class t
 template<typename... Ts> requires
 false_v<requires (Ts... ts) {requires ((sizeof(ts) == 2) && ...);}>
 // expected-note@-1 {{because 'false_v<requires (short ts, unsigned short ts) { requires (sizeof (ts) == 2) && (sizeof (ts) == 2); }>'}}
-// expected-note@-2 {{because 'false_v<requires (short ts) { requires (sizeof (ts) == 2); }>' evaluated to false}}
+// expected-note@-2 {{because 'false_v<requires (short ts) { requires sizeof (ts) == 2; }>' evaluated to false}}
```

Comment:
Hmm... this is the only one that causes a touch of heartburn, but I don't really know how I feel/if I care enough.  it would be interesting to see if @cor3ntin has an opinion here.

Reply:
Yeah... I don't think we should modify the AST to avoid a warning. Can we instead tweak when the warning is emitted?

Reply:
Thats the discussion we had in a separate thread, we don't really see how, without modifying the ParenExpr to say that it is the result of a fold expr expansion.

As far as 'modify the ast', depending on your perspective/mental model here, the parens are a part of the 'fold expr' rather than part of the expr being folded.  In this case, if you consider the `( ) && ...` to effectively be the fold expr syntax, removing it seems to make a ton of sense.  

But yeah, we couldn't think of a good way to suppress the warning otherwise (or, like i mentioned, a flag on `ParenExpr` to say that the `ParenExpr` is a part of a single-element-fold-expr-expansion).  The diagnostic just sees the expression after expansion, so as far as it knows, you did `if ((a == b))` instead of `if ((a==b) && ...)` but only had a single element.

Reply:
*NOTE: The mental model of 'parens are a necessary part of the fold expr' is not in the standard of course, which just uses a 'cast-expr' for that, but is it effectively so?

Reply:
> Thats the discussion we had in a separate thread, we don't really see how, without modifying the ParenExpr to say that it is the result of a fold expr expansion.

Honestly not the worst idea?

We should test that the following still works

```cpp
void f(auto a, auto... b) {
    static_assert(__is_same_as(decltype((a)), int&));
    static_assert(__is_same_as(decltype(((b), ...)), int&));
};

int main() {
    int a = 0;
    f(a, a);
}
```

Reply:
So there is no reason that shouldn't still work fine, right?  The AST is already 'set', other than the instantiated items, so no 'reparsing' can happen where the paren expr would matter, as the decltype would already be set as to whether it is a expr or type based variant, right?  Though it is the one I'm least sure about.

The only thing I could come up with that this MIGHT cause an issue (which, not ruling out the Decltype above!), is a case where we do a ast-print, and re-parse.  But this requires our ast-print to produce 'compile-able code', which isn't necessarily the case anyway.  Also, to re-parse somehow a template instantiation.

THOUGH, looking back at `ActOnDecltypeExpression` (called when transforming the `DecltypeType`, there IS a check for the `ParenExpr`, so I think you've found a reason we shouldn't do this.  It seems wasteful to make `ParenExpr` store a bit for this just for hte diagnostic, but hopefully this will be useful for something else in the future, but I think we HAVE to.

Reply:
> The only thing I could come up with that this MIGHT cause an issue (which, not ruling out the Decltype above!),

I briefly looked into some new cases with `decltype`, and it seems it's unsafe to omit the `()`. I’ll take a closer look at those cases...